File: code/packages/ruby/riscv_simulator/lib/coding_adventures/riscv_simulator/opcodes.rb
# frozen_string_literal: true

# === RISC-V RV32I opcode constants ===
#
# Every RISC-V instruction is exactly 32 bits wide. The lowest 7 bits ([6:0])
# always contain the "opcode" -- which tells the CPU what category of work
# this instruction does. Within each category, funct3 and funct7 fields
# narrow down the exact operation.

module CodingAdventures
  module RiscvSimulator
    # === RV32I Base Integer Opcodes ===
    OPCODE_LOAD    = 0b0000011
    OPCODE_STORE   = 0b0100011
    OPCODE_BRANCH  = 0b1100011
    OPCODE_JAL     = 0b1101111
    OPCODE_JALR    = 0b1100111
    OPCODE_LUI     = 0b0110111
    OPCODE_AUIPC   = 0b0010111
    OPCODE_OP_IMM  = 0b0010011
    OPCODE_OP      = 0b0110011
    OPCODE_SYSTEM  = 0b1110011

    # === Funct3 for I-type immediate arithmetic ===
    FUNCT3_ADDI  = 0
    FUNCT3_SLTI  = 2
    FUNCT3_SLTIU = 3
    FUNCT3_XORI  = 4
    FUNCT3_ORI   = 6
    FUNCT3_ANDI  = 7
    FUNCT3_SLLI  = 1
    FUNCT3_SRLI  = 5 # also SRAI

    # === Funct3 for R-type arithmetic ===
    FUNCT3_ADD  = 0 # also SUB
    FUNCT3_SLL  = 1
    FUNCT3_SLT  = 2
    FUNCT3_SLTU = 3
    FUNCT3_XOR  = 4
    FUNCT3_SRL  = 5 # also SRA
    FUNCT3_OR   = 6
    FUNCT3_AND  = 7

    # === Funct7 ===
    FUNCT7_NORMAL = 0x00
    FUNCT7_ALT    = 0x20

    # === Funct3 for loads ===
    FUNCT3_LB  = 0
    FUNCT3_LH  = 1
    FUNCT3_LW  = 2
    FUNCT3_LBU = 4
    FUNCT3_LHU = 5

    # === Funct3 for stores ===
    FUNCT3_SB = 0
    FUNCT3_SH = 1
    FUNCT3_SW = 2

    # === Funct3 for branches ===
    FUNCT3_BEQ  = 0
    FUNCT3_BNE  = 1
    FUNCT3_BLT  = 4
    FUNCT3_BGE  = 5
    FUNCT3_BLTU = 6
    FUNCT3_BGEU = 7

    # === Funct3 for system ===
    FUNCT3_PRIV  = 0
    FUNCT3_CSRRW = 1
    FUNCT3_CSRRS = 2
    FUNCT3_CSRRC = 3

    # === Funct7 for privileged instructions ===
    FUNCT7_ECALL = 0x00
    FUNCT7_MRET  = 0x18
  end
end

File: code/packages/rust/riscv-simulator/src/opcodes.rs
//! Opcode constants for the RV32I base integer ISA.

// RV32I Base Integer Opcodes
pub const OPCODE_LOAD: u32    = 0b0000011;
pub const OPCODE_STORE: u32   = 0b0100011;
pub const OPCODE_BRANCH: u32  = 0b1100011;
pub const OPCODE_JAL: u32     = 0b1101111;
pub const OPCODE_JALR: u32    = 0b1100111;
pub const OPCODE_LUI: u32     = 0b0110111;
pub const OPCODE_AUIPC: u32   = 0b0010111;
pub const OPCODE_OP_IMM: u32  = 0b0010011;
pub const OPCODE_OP: u32      = 0b0110011;
pub const OPCODE_SYSTEM: u32  = 0b1110011;

// Funct3 for I-type immediate arithmetic
pub const FUNCT3_ADDI: u32  = 0;
pub const FUNCT3_SLTI: u32  = 2;
pub const FUNCT3_SLTIU: u32 = 3;
pub const FUNCT3_XORI: u32  = 4;
pub const FUNCT3_ORI: u32   = 6;
pub const FUNCT3_ANDI: u32  = 7;
pub const FUNCT3_SLLI: u32  = 1;
pub const FUNCT3_SRLI: u32  = 5;

// Funct3 for R-type arithmetic
pub const FUNCT3_ADD: u32  = 0;
pub const FUNCT3_SLL: u32  = 1;
pub const FUNCT3_SLT: u32  = 2;
pub const FUNCT3_SLTU: u32 = 3;
pub const FUNCT3_XOR: u32  = 4;
pub const FUNCT3_SRL: u32  = 5;
pub const FUNCT3_OR: u32   = 6;
pub const FUNCT3_AND: u32  = 7;

// Funct7
pub const FUNCT7_NORMAL: u32 = 0x00;
pub const FUNCT7_ALT: u32    = 0x20;

// Funct3 for loads
pub const FUNCT3_LB: u32  = 0;
pub const FUNCT3_LH: u32  = 1;
pub const FUNCT3_LW: u32  = 2;
pub const FUNCT3_LBU: u32 = 4;
pub const FUNCT3_LHU: u32 = 5;

// Funct3 for stores
pub const FUNCT3_SB: u32 = 0;
pub const FUNCT3_SH: u32 = 1;
pub const FUNCT3_SW: u32 = 2;

// Funct3 for branches
pub const FUNCT3_BEQ: u32  = 0;
pub const FUNCT3_BNE: u32  = 1;
pub const FUNCT3_BLT: u32  = 4;
pub const FUNCT3_BGE: u32  = 5;
pub const FUNCT3_BLTU: u32 = 6;
pub const FUNCT3_BGEU: u32 = 7;

// Funct3 for system
pub const FUNCT3_PRIV: u32  = 0;
pub const FUNCT3_CSRRW: u32 = 1;
pub const FUNCT3_CSRRS: u32 = 2;
pub const FUNCT3_CSRRC: u32 = 3;

// Funct7 for privileged instructions
pub const FUNCT7_ECALL: u32 = 0x00;
pub const FUNCT7_MRET: u32  = 0x18;
